packages/design-system/src/components/analytics/getAnalyticsContentFromRefs.ts

```diff
@@ -1,7 +1,34 @@
 import { RefObject } from 'react';
 
+/**
+ * Gets text content from an element and even follows `slot` elements across Shadow DOM
+ * barriers to get the full text that the user sees.
+ */
+function getTextContent(nodes: Node[]): string {
+  return nodes
+    .map((node) => {
+      if (node.nodeType === Node.TEXT_NODE) {
+        return node.textContent;
+      } else if (node.nodeType === Node.ELEMENT_NODE) {
```

Comment:
@pwolfert  thanks for the accompanying video walking through this—it helped me a lot. I was wondering, for readability, if there might be a way to make the logic a bit more modular since there's a lot going on with the recursion and inline branching?

Reply:
I'm not really sure what you mean. Can you give an example of how you might refactor it?

Reply:
Sure! My idea is to break out the logic that handles the element node cases into separate functions. Here's a rough sketch of what I mean—let me know what you think!
```
function getTextContentFromSlot(node: HTMLSlotElement): string {
  const assignedNodes = node.assignedNodes({ flatten: true });
  return getTextContent(assignedNodes);
}

function getTextContentFromElementNode(node: HTMLElement): string {
  if (node.tagName === 'SLOT') {
    // If it's a slot, it won't actually have any content in itself, so we have to
    // follow the trail to its "assigned nodes" and look there.
    return getTextContentFromSlot(node as HTMLSlotElement);
  }
  return getTextContent(Array.from(node.childNodes));
}

function processNode(node: Node): string {
  if (node.nodeType === Node.TEXT_NODE) {
    return node.textContent;
  } else if (node.nodeType === Node.ELEMENT_NODE) {
    return getTextContentFromElementNode(node as HTMLElement);
  } 

  return '';
}

function getTextContent(nodes: Node[]): string {
  return nodes.map(processNode).join('');
}
```

Reply:
I see what you mean now! Since none of the functions are meant to be used on their own, I personally prefer having all the logic contained in that one recursive function. Do you feel strongly about it?

Reply:
No, not at all. Your reasoning makes sense to me